linux/ibus-keyman/src/engine.c

```diff
@@ -241,40 +241,39 @@ client_supports_surrounding_text(IBusEngine *engine) {
 
 static void
 reset_context(IBusEngine *engine) {
-    IBusKeymanEngine *keyman = (IBusKeymanEngine *)engine;
+  IBusKeymanEngine *keyman = (IBusKeymanEngine *)engine;
+  km_kbp_context *context;
+
+  g_message("%s", __FUNCTION__);
+  context = km_kbp_state_context(keyman->state);
+
+  if (client_supports_surrounding_text(engine)) {
     IBusText *text;
     gchar *surrounding_text, *current_context_utf8;
-    guint cursor_pos, anchor_pos, context_start, context_pos;
+    guint cursor_pos, anchor_pos, context_start, context_end;
     km_kbp_context_item *context_items;
-    km_kbp_context *context;
 
-    g_message("reset_context");
-    context = km_kbp_state_context(keyman->state);
-    if ((engine->client_capabilities & IBUS_CAP_SURROUNDING_TEXT) != 0)
-    {
-        current_context_utf8 = get_current_context_text(context);
-
-        ibus_engine_get_surrounding_text(engine, &text, &cursor_pos, &anchor_pos);
-        context_pos = anchor_pos < cursor_pos ? anchor_pos : cursor_pos;
-        context_start = context_pos > MAXCONTEXT_ITEMS ? context_pos - MAXCONTEXT_ITEMS : 0;
-        surrounding_text = g_utf8_substring(ibus_text_get_text(text), context_start, context_pos);
-        g_message("new context is:%u:%s: cursor:%d anchor:%d", context_pos - context_start, surrounding_text, cursor_pos, anchor_pos);
-
-        g_message(":%s:%s:", surrounding_text, current_context_utf8);
-        if (!g_str_has_suffix(surrounding_text, current_context_utf8))
-        {
-            g_message("setting context because it has changed from expected");
-            if (km_kbp_context_items_from_utf8(surrounding_text, &context_items) == KM_KBP_STATUS_OK) {
-                km_kbp_context_set(context, context_items);
-            }
-            km_kbp_context_items_dispose(context_items);
-        }
-        g_free(surrounding_text);
-        g_free(current_context_utf8);
-    }
-    else {
-        km_kbp_context_clear(context);
+    ibus_engine_get_surrounding_text(engine, &text, &cursor_pos, &anchor_pos);
+
+    context_end      = anchor_pos < cursor_pos ? anchor_pos : cursor_pos;
+    context_start    = context_end > MAXCONTEXT_ITEMS ? context_end - MAXCONTEXT_ITEMS : 0;
+    surrounding_text = g_utf8_substring(ibus_text_get_text(text), context_start, context_end);
+    g_message("%s: new context is :%s: (len:%u) cursor:%d anchor:%d", __FUNCTION__,
+      surrounding_text, context_end - context_start, cursor_pos, anchor_pos);
+
+    current_context_utf8 = get_current_context_text(context);
+    if (!g_str_has_suffix(surrounding_text, current_context_utf8) || !g_utf8_strlen(current_context_utf8, -1)) {
+      g_message("%s: setting context because it has changed from expected", __FUNCTION__);
+      if (km_kbp_context_items_from_utf8(surrounding_text, &context_items) == KM_KBP_STATUS_OK) {
+        km_kbp_context_set(context, context_items);
+        km_kbp_context_items_dispose(context_items);
+      }
```

Comment:
If this fails, I think we should `km_kbp_context_clear(context)` and log.
```suggestion
      if ((status = km_kbp_context_items_from_utf8(surrounding_text, &context_items)) == KM_KBP_STATUS_OK) {
        km_kbp_context_set(context, context_items);
        km_kbp_context_items_dispose(context_items);
      } else {
        km_kbp_context_clear(context);
        g_message("%s: setting context failed with %d", __FUNCTION__, status);
      }
```

Reply:
See #7540.